### PR TITLE
mgr: add volumes module

### PR DESCRIPTION
**Pull request by liewegas**
This is currently leaving ceph_volume_client.py in its current location and re-using it, but we can later move that into the module, and replace ceph_volume_client with a stub that calls out to the mgr.

Simplified version of #23531 (removed dependency on poolsets, and cherry-picked a few commits from that branch).

### CODE REVIEW

File: src/pybind/mgr/volumes/module.py [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+        #   skip purge jobs if their fscid no longer exists
+
+        # TODO: on volume delete, cancel out any background jobs that
+        # affect subvolumes within that volume.
+
+        # ... any background init needed?  Can get rid of this
+        # and _initialized if not
+        self._initialized.set()
+
+    def handle_command(self, inbuf, cmd):
```
Comment (sebastian-philipp) [inline src/pybind/mgr/volumes/module.py]: I'd be :+1:  for making this the default implementation for all modules

File: src/pybind/mgr/volumes/module.py [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+
+        handler_name = "_cmd_" + cmd['prefix'].replace(" ", "_")
+        try:
+            handler = getattr(self, handler_name)
+        except AttributeError:
+            return -errno.EINVAL, "", "Unknown command"
+
+        return handler(inbuf, cmd)
+
+    def _orchestrator_wait(self, completions):
```
Comment (sebastian-philipp) [inline src/pybind/mgr/volumes/module.py]: This seems to be an outdated copy of `orchestrator_cli.module.OrchestratorCli#_wait`. Maybe, we should unify this into a single `OrchestratorClient` class.

File: src/pybind/mgr/volumes/module.py [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+        self._background_jobs = Queue.Queue()
+
+    def _oremote(self, *args, **kwargs):
+        """
+        Helper for invoking `remote` on whichever orchestrator is enabled
+        """
+        # TODO: orchestrator_cli has a setting for orchestrator backend,
+        # but we don't really want to start propagating that into
+        # other modules individually -- maybe mgrmap should have
+        # the setting?  or just a C++-level mgr config option?
```
Comment (sebastian-philipp) [inline src/pybind/mgr/volumes/module.py]: Added to the orchestrator TODO list.

File: src/pybind/mgr/mgr_module.py [hunk 4]
```diff
@@ -566,6 +568,28 @@ def get_daemon_status(self, svc_type, svc_id):
         """
         return self._ceph_get_daemon_status(svc_type, svc_id)
 
+    def mon_command(self, cmd_dict):
```
Comment (sebastian-philipp) [inline src/pybind/mgr/mgr_module.py]: Hm. Looks very similar to `dashboard.services.ceph_service.CephService#send_command`

File: src/mon/MgrMonitor.cc [hunk 5]
```diff
@@ -44,7 +44,8 @@ const static std::map<uint32_t, std::set<std::string>> always_on_modules = {
       "crash",
       "status",
       "balancer",
-      "devicehealth"
+      "devicehealth",
+      "volumes",
```
Comment (jtlayton) [inline src/mon/MgrMonitor.cc]: Since volumes doesn't exist yet at this point, shouldn't this be the last patch in the series?

**Comment by liewegas** [comment]
@jan--f this definitely is intended to be a general cephfs thing, which manila as just the first user.  ceph-csi for kubernetes will immediately follow

File: src/pybind/mgr/volumes/module.py (hunk 2 above)
Comment (liewegas) [reply src/pybind/mgr/volumes/module.py]: Sounds good to me... can that be a follow-on cleanup?

**Comment by liewegas** [comment]
Reordered commits, fixed vstart.sh to use the new module.  Look ok?

**Comment by liewegas** [comment]
also dropped the unrelated orchestrator non-ascii cleanup

**Comment by liewegas** [comment]
@batrick see 43bf12e

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2018-11-28_15:02:32-fs:basic_functional-wip-sage2-testing-2018-11-27-1650-distro-basic-smithi/

**Comment by liewegas** [comment]
retest this please

**Pull request merged**